Meta/run.py

```diff
@@ -0,0 +1,887 @@
+
+from __future__ import annotations
+
+import os
+import re
+import sys
+from dataclasses import dataclass, field
+from enum import Enum, unique
+from itertools import chain, repeat
+from os import access, environ
+from pathlib import Path
+from shutil import which
+from subprocess import run
+from typing import Any, Callable, Literal
+
+QEMU_MINIMUM_REQUIRED_MAJOR_VERSION = 6
+QEMU_MINIMUM_REQUIRED_MINOR_VERSION = 2
+
+
+class RunError(Exception):
+    pass
+
+
+@unique
+class Arch(Enum):
+    """SerenityOS architecture, not host architecture."""
+
+    Aarch64 = "aarch64"
+    x86_64 = "x86_64"
+
+
+@unique
+class QEMUKind(Enum):
+    """VM distrinctions determining which virtualizers *might* be used."""
+
+    # Linux and anything that may or may not have KVM, including WSL with Linux QEMU.
+    Other = "kvm"
+    # WSL with native Windows QEMU (and possibly WHPX).
+    NativeWindows = "whpx"
+    # MacOS with HVF.
+    MacOS = "hvf"
+    # Serenity on Serenity with ported QEMU
+    SerenityOS = "tcg"
+
+
+@unique
+class MachineType(Enum):
+    Default = ""
+    QEMUTap = "qtap"
+    QEMUWithoutNetwork = "qn"
+    QEMUExtLinux = "qextlinux"
+    QEMU35 = "q35"
+    QEMU35Grub = "q35grub"
+    QEMUGrub = "qgrub"
+    CI = "ci"
+    Limine = "limine"
+    Bochs = "b"
+    MicroVM = "microvm"
+    ISAPC = "isapc"
+
+    def uses_grub(self) -> bool:
+        return self in [MachineType.QEMU35Grub, MachineType.QEMUGrub]
+
+    def is_q35(self) -> bool:
+        return self in [MachineType.QEMU35Grub, MachineType.QEMU35]
+
+    def supports_pc_speaker(self) -> bool:
+        """Whether the pcspk-audiodev option is allowed for this machine type."""
+        return self in [
+            MachineType.Default,
+            MachineType.QEMUTap,
+            MachineType.QEMUWithoutNetwork,
+            MachineType.QEMUExtLinux,
+            MachineType.QEMU35,
+            MachineType.QEMU35Grub,
+            MachineType.QEMUGrub,
+            MachineType.CI,
+            MachineType.Limine,
+        ]
+
+
+@dataclass
+class Configuration:
+    """Run configuration, populated from command-line or environment variable data."""
+
+    # ## Programs and environmental configuration
+    virtualization_support: bool = False
+    bochs_binary: Path = Path("bochs")
+    qemu_binary: Path | None = None
+    qemu_kind: QEMUKind | None = None
+    kvm_usable: bool | None = None
+    architecture: Arch | None = None
+    serenity_src: Path | None = None
+
+    # ## High-level run configuration
+    machine_type: MachineType = MachineType.Default
+    enable_gdb: bool = False
+    enable_gl: bool = False
+    nvme_enable: bool = True
+    sd_enable: bool = False
+    screen_count: int = 1
+    host_ip: str = "127.0.0.1"
+    ethernet_device_type: str = "e1000"
+    disk_image: Path = Path("_disk_image")
+
+    # ## Low-level QEMU configuration
+    # QEMU -append
+    kernel_cmdline: list[str] = field(default_factory=lambda: ["hello"])
```

Comment:
Since the list is constructed at the time the class definition is executed (not when the class is constructed), all instances would share the same list. This is the same reason that lists cannot be used as default arguments in functions directly.

Reply:
But, we only have one instance of `Configuration` as far as I can tell, so that shouldn't matter.

PS: Please don't mark comments as resolved if they aren't resolved. That just makes going back to respond annoying.

Reply:
This is idiomatic and recommended by the dataclass documentation, [see here](https://docs.python.org/3/library/dataclasses.html#dataclasses.field).

Reply:
I'm sure Python considers this idiomatic. But if we don't need it, I'd argue we should keep it more readable.